onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -2073,22 +2073,23 @@ static Status RegisterCudaKernels(KernelRegistry& kernel_registry) {
 
 }  // namespace cuda
 
-static std::shared_ptr<onnxruntime::KernelRegistry> s_kernel_registry;
+static std::shared_ptr<KernelRegistry>& CudaKernelRegistry() {
+  // static local variable ensures thread-safe initialization
+  static std::shared_ptr<KernelRegistry> cuda_kernel_registry = []() {
+    std::shared_ptr<KernelRegistry> registry = KernelRegistry::Create();
+    ORT_THROW_IF_ERROR(cuda::RegisterCudaKernels(*registry));
+    return registry;
+  }();
+
+  return cuda_kernel_registry;
+}
 
 void Shutdown_DeleteRegistry() {
-  s_kernel_registry.reset();
+  CudaKernelRegistry().reset();
 }
 
 std::shared_ptr<KernelRegistry> CUDAExecutionProvider::GetKernelRegistry() const {
-  if (!s_kernel_registry) {
-    s_kernel_registry = KernelRegistry::Create();
-    auto status = cuda::RegisterCudaKernels(*s_kernel_registry);
-    if (!status.IsOK())
-      s_kernel_registry.reset();
-    ORT_THROW_IF_ERROR(status);
-  }
-
-  return s_kernel_registry;
+  return CudaKernelRegistry();
```

Comment:
Is there any advantage of having an additional helper function here?

Reply:
Here, the helper's result (shared_ptr&) is copied. But the helper returns a reference so its shared_ptr can be reset by Shutdown_DeleteRegistry().

Reply:
Ah, didn't see the reference.